sdk/dotnet/Pulumi/Core/Input.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Internal interface to allow our code to operate on inputs in an untyped manner. Necessary as
+    /// there is no reasonable way to write algorithms over heterogeneous instantiations of generic
+    /// types.
+    /// </summary>
+    internal interface IInput
+    {
+        IOutput ToOutput();
+    }
+
+    /// <summary>
+    /// <see cref="Input{T}"/> is a property input for a <see cref="Resource"/>.  It may be a promptly
+    /// available T, or the output from a existing <see cref="Resource"/>.
+    /// </summary>
+    public class Input<T> : IInput
+    {
+        /// <summary>
+        /// Technically, in .net we can represent Inputs entirely using the Output type (since
+        /// Outputs can wrap values and promises).  However, it would look very weird to state that
+        /// the inputs to a resource *had* to be Outputs. So we basically just come up with this
+        /// wrapper type so things look sensible, even though under the covers we implement things
+        /// using the exact same type

```

Comment:
> Should we then consider renaming Output to something more suitable and killing Input?

I would really like to *not* do that.  I gave this a bunch of a thought and that would cause our APIs to start splitting between platforms.  i.e. we talk aobut 'inputs' and 'outputs' all over the place.  We'd need to start treating .NET specially and explaining: ok, when our docs say X, it means Y over in .NET.

I think consistency is important here, if we want to keep costs under control.

Reply:
Agreed.  Going to keep Input/Output.